Game/data/building_systems/heat_system/heat_flow_graph.gd

```diff
@@ -0,0 +1,156 @@
+## The vertices of the graph are stored as building positions on the grid, as that makes it easier
+## to figure out adjacencies.
+var graph: DirectedWeightedGraph
+
+## Store a map to convert vertex positions to their corresponding heat components.
+var vertex_to_component_map: Dictionary[Vector2i, HeatComponent]
+
+
+## Create a new heat flow graph containing only the omni-source and omni-sink.
+func _init() -> void:
+	graph = DirectedWeightedGraph.new()
+	graph.add_vertex(SOURCE)
+	graph.add_vertex(SINK)
+
+
+## Given a position, give the 4 positions that are cardinally adjacent to that position.
+## Used internally to figure out adjacencies.
+func _get_neighbors(position: Vector2i):
+	return [
+		position + Vector2i.RIGHT,
+		position + Vector2i.UP,
+		position + Vector2i.LEFT,
+		position + Vector2i.DOWN,
+	]
+
+
+## Add one edge with start -> end with the given weight, and another edge end -> start with weight
+## 0. This is needed to make running Ford-Fulkerson easier.
+func _add_two_way_flow_edge(start, end, weight) -> void:
+	graph.add_edge(start, end, weight)
+	graph.add_edge(end, start, 0)
+
+
+## Given a heat component, add that building to the heat flow graph.
+## If it is a source or sink, connect it to the omni-source/omni-sink.
+## If it is adjacent to any buildings already in the heat flow graph,
+## add edges flowing out of the source or into the sink.
+func add_building(heat_component: HeatComponent) -> void:
+	var position: Vector2i = heat_component.building_entity.position
+	graph.add_vertex(position)
+	vertex_to_component_map[position] = heat_component
+
+	if heat_component.is_source:
+		# omni-source -> source
+		_add_two_way_flow_edge(SOURCE, position, heat_component.heat_production)
+		for neighbor_position: Vector2i in _get_neighbors(position):
+			if graph.has_vertex(neighbor_position):
+				# source -> neighbors
+				_add_two_way_flow_edge(position, neighbor_position, HEAT_MAX_FLOW)
+
+	elif heat_component.is_sink:
+		# sink -> omni-sink
+		_add_two_way_flow_edge(position, SINK, heat_component.heat_passive_cool_off)
+		for neighbor_position: Vector2i in _get_neighbors(position):
+			if graph.has_vertex(neighbor_position):
+				# neighbors -> sink
+				_add_two_way_flow_edge(neighbor_position, position, HEAT_MAX_FLOW)
+
+	else:
+		assert(false, "CARRIERS not implemented yet")
+
+
+## Given a heat component, remove that building from the heat flow graph.
+func remove_building(heat_component: HeatComponent) -> void:
+	var position: Vector2i = heat_component.building_entity.position
+	graph.remove_vertex(position)
+	vertex_to_component_map.erase(position)
+
+
+## Given a vertex in the graph, get its corresponding heat component.
+## Return null if there isn't one.
+func get_component_at(position: Vector2i) -> HeatComponent:
+	return vertex_to_component_map.get(position)
+
+
+## Do a deep copy of this object and return the duplicate.
+func duplicate_graph() -> HeatFlowGraph:
```

Comment:
consider creating a utility to generically do a deep copy of objects. Or use one of the existing utilities built for this purpose, such as Object.duplicate.